source/sigv4.c

```diff
@@ -436,6 +561,402 @@ static SigV4Status_t parseDate( const char * pDate,
+                                         SigV4String_t * pCredScope )
+{
+    SigV4Status_t returnVal = SigV4InvalidParameter;
+    char * pBufWrite = NULL;
+    int32_t bytesWritten = 0;
+
+    assert( pSigV4Params != NULL );
+    assert( pCredScope != NULL );
+    assert( pCredScope->pData != NULL );
+
+    pBufWrite = pCredScope->pData;
+
+    /* Use only the first 8 characters from the provided ISO 8601 string (YYYYMMDD). */
+    bytesWritten = snprintf( ( char * ) pBufWrite,
+                             ISO_DATE_SCOPE_LEN + 1,
+                             "%*s",
+                             ISO_DATE_SCOPE_LEN,
+                             pSigV4Params->pDateIso8601 );
+
+    if( bytesWritten == ISO_DATE_SCOPE_LEN )
+    {
+        bytesWritten = snprintf( ( char * ) pBufWrite,
+                                 pSigV4Params->regionLen,
+                                 "/%s/",
+                                 pSigV4Params->pRegion );
+
+        if( bytesWritten != pSigV4Params->regionLen )
+        {
+            LogError( ( "Error in formatting provided region string for credential scope." ) );
+            returnVal = SigV4ISOFormattingError;
+        }
+    }
+    else
+    {
+        LogError( ( "Error obtaining date for credential scope string." ) );
+        returnVal = SigV4ISOFormattingError;
+    }
+
+    if( returnVal != SigV4ISOFormattingError )
+    {
+        bytesWritten = snprintf( ( char * ) pBufWrite,
+                                 pSigV4Params->serviceLen,
+                                 "/%s/",
+                                 pSigV4Params->pService );
+
+        if( bytesWritten != pSigV4Params->serviceLen )
+        {
+            LogError( ( "Error in formatting provided service string for credential scope." ) );
+            returnVal = SigV4ISOFormattingError;
+        }
+    }
+    else
+    {
+        pCredScope->dataLen = ISO_DATE_SCOPE_LEN + pSigV4Params->regionLen + pSigV4Params->serviceLen;
+        returnVal = SigV4Success;
+    }
+
+    return returnVal;
+}
+
+/*-----------------------------------------------------------*/
+
+#if ( SIGV4_USE_CANONICAL_SUPPORT == 1 )
+
+    static void stringToSigV4Value( SigV4String_t * pSigV4Value,
+                                    const char * pInput,
+                                    size_t lenInput )
+    {
+        assert( pSigV4Value != NULL );
+        assert( pInput != NULL );
+        assert( lenInput > 0U );
+
+        pSigV4Value->pData = ( char * ) pInput;
+        pSigV4Value->dataLen = ( size_t ) lenInput;
+    }
+
+/*-----------------------------------------------------------*/
+
+    static bool emptySigV4String( SigV4String_t * pInput )
+    {
+        bool returnVal = true;
+
+        assert( pInput != NULL );
+
+        return ( pInput->pData == NULL || pInput->dataLen == 0 ) ? returnVal : !returnVal;
+    }
+
+/*-----------------------------------------------------------*/
+
+    static int cmpKeyValue( const void * pFirstVal,
```

Comment:
why use `void*` here? Can we not use `SigV4KeyValuePair_t`?

Reply:
It was just to match the function signature that's defined for the `qsort()` comparison functions, which is the intention for this function as well (during canonicalization).